datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -290,6 +293,84 @@ fn filter_array(
     filter.filter(array)
 }
 
+fn merge(
```

Comment:
could it be called maybe `zip` or `zip_columnar_value` ?

Reply:
It's almost the same as `zip`, but different enough that it's necessary. Without this implementation you can't avoid the scatter step.

I've added a test case to show the difference. The short version is that `merge([true, false, true], [A, C], [B])` will get you `[A, B, C]` while `zip` would return an error stating `all arrays should have the same length`.

I agree that these two merge kernels would be better off in `arrow-rs` which is why I made PR https://github.com/apache/arrow-rs/pull/8753.